pki.go

```diff
@@ -0,0 +1,250 @@
+package nebula
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/slackhq/nebula/cert"
+	"github.com/slackhq/nebula/config"
+	"github.com/slackhq/nebula/util"
+)
+
+type PKI struct {
+	cs     atomic.Pointer[CertState]
+	caPool atomic.Pointer[cert.NebulaCAPool]
+	l      *logrus.Logger
+}
+
+type CertState struct {
+	Certificate         *cert.NebulaCertificate
+	RawCertificate      []byte
+	RawCertificateNoKey []byte
+	PublicKey           []byte
+	PrivateKey          []byte
+}
+
+func NewPKIFromConfig(l *logrus.Logger, c *config.C) (*PKI, error) {
+	pki := &PKI{l: l}
+	err := pki.reload(c, true)
+	if err != nil {
+		l.WithError(err).Error("WHAT THE HELL")
+		return nil, err
+	}
+
+	c.RegisterReloadCallback(func(c *config.C) {
+		cErr := pki.reload(c, false)
+		if cErr != nil {
+			cErr.Log(l)
+		}
+	})
+
+	l.Error("SO NO ERROR THEN")
+	return pki, nil
+}
+
+func (p *PKI) GetCertState() *CertState {
+	return p.cs.Load()
+}
+
+func (p *PKI) GetCAPool() *cert.NebulaCAPool {
+	return p.caPool.Load()
+}
+
+func (p *PKI) reload(c *config.C, initial bool) *util.ContextualError {
```

Comment:
Usually its a bad pattern to return anything but the `error` interface or weird things can happen with non-nil types and stuff if something tries to wrap this as an `error` later.

https://go.dev/doc/faq#nil_error

> It's a good idea for functions that return errors always to use the error type in their signature (as we did above) rather than a concrete type such as *MyError, to help guarantee the error is created correctly. As an example, [os.Open](https://go.dev/pkg/os/#Open) returns an error even though, if not nil, it's always of concrete type [*os.PathError](https://go.dev/pkg/os/#PathError).

It's probably okay since these methods are private and internal, but it can cause really hard to figure out bugs

Reply:
Totally agree, just kind of lost on the approach we should take going forward.

We are moving towards this pattern where a struct can initialize itself from config and can handle its own reload using mostly the same code path, I think its a good direction to move in. But it creates some awkwardness with error handling and we've solved these situation in the past with `util.ContextualError`. 

if we return `error` instead of `*util.ContextualError` then we need to check each error return where we may have a `util.ContextualError` to avoid adding further context and making the log line silly.

Basically I am hoping to avoid this becoming pervasive in `main` https://github.com/slackhq/nebula/blob/master/main.go#L225-L230

Instead of returning the error I guess we could handle logging in the `New*` functions and return an `ok`, thoughts?

Reply:
@wadey had the genius idea of just using helper functions so I moved to that.